arch/x86/core/early_serial.c

```diff
@@ -1,102 +1,67 @@
-#define UART_REG_ADDR_INTERVAL \
-	(1 << DT_PROP(DT_CHOSEN(zephyr_console), reg_shift))
-#endif
-
+/* Legacy I/O Port Access to a NS16550 UART */
 #ifdef UART_NS16550_ACCESS_IOPORT
-#define PORT		((io_port_t)DT_REG_ADDR(DT_CHOSEN(zephyr_console)))
-#define INBYTE(x)	sys_in8(x)
-#define INWORD(x)	sys_in32(x)
-#define OUTBYTE(x, d)	sys_out8(d, x)
-#define OUTWORD(x, d)	sys_out32(d, x)
-#ifndef UART_REG_ADDR_INTERVAL
-#define UART_REG_ADDR_INTERVAL 1 /* address diff of adjacent regs. */
-#endif /* UART_REG_ADDR_INTERVAL */
-#else
-#define PORT		((mm_reg_t)DT_REG_ADDR(DT_CHOSEN(zephyr_console)))
-#define INBYTE(x)	sys_read8(x)
-#define INWORD(x)	sys_read32(x)
-#define OUTBYTE(x, d)	sys_write8(d, x)
-#define OUTWORD(x, d)	sys_write32(d, x)
-#ifndef UART_REG_ADDR_INTERVAL
-#define UART_REG_ADDR_INTERVAL 4 /* address diff of adjacent regs. */
+#define IN(reg)       sys_in8(reg)
+#define OUT(reg, val) sys_out8(val, reg)
 #endif
-#endif /* UART_NS16550_ACCESS_IOPORT */
 
-#ifdef CONFIG_UART_NS16550_ACCESS_WORD_ONLY
-#undef INBYTE
-#define INBYTE(x) INWORD(x)
-#undef OUTBYTE
-#define OUTBYTE(x, d) OUTWORD(x, d)
+/* "Modern" mapping of a UART into a PCI MMIO device.  The registers
+ * are still bytes, but spaced at a 32 bit stride instead of packed
+ * together.
+ */
+#ifdef X86_SOC_EARLY_SERIAL_PCIDEV
+static uintptr_t pci_bar;
+#define IN(reg)       (sys_read32(pci_bar + reg * 4) & 0xff)
+#define OUT(reg, val) sys_write32((val) & 0xff, pci_bar + reg * 4)
 #endif
 
-#ifdef UART_NS16550_PCIE_ENABLED
-static mm_reg_t base;
-#else
-#define base PORT
+/* Still other devices use a MMIO region containing packed byte
+ * registers
+ */
+#ifdef X86_SOC_EARLY_SERIAL_MMIO8_ADDR
+#define IN(reg)       sys_read8(X86_SOC_EARLY_SERIAL_MMIO8_ADDR + reg)
+#define OUT(reg, val) sys_write8(val, X86_SOC_EARLY_SERIAL_MMIO8_ADDR + reg)
 #endif
 
+#define REG_THR  0x00  /* Transmitter holding reg. */
+#define REG_IER  0x01  /* Interrupt enable reg.    */
+#define REG_FCR  0x02  /* FIFO control reg.        */
+#define REG_LCR  0x03  /* Line control reg.        */
+#define REG_MCR  0x04  /* Modem control reg.       */
+#define REG_LSR  0x05  /* Line status reg.         */
+#define REG_BRDL 0x00  /* Baud rate divisor (LSB)  */
+#define REG_BRDH 0x01  /* Baud rate divisor (MSB)  */
+
+#define IER_DISABLE     0x00
+#define LCR_8N1         (BIT(0) | BIT(1))
+#define LCR_DLAB_SELECT BIT(7)
+#define MCR_DTR         BIT(0)
+#define MCR_RTS         BIT(1)
+#define LSR_THRE        BIT(5)
+
+#define FCR_FIFO    BIT(0)  /* enable XMIT and RCVR FIFO */
+#define FCR_RCVRCLR BIT(1)  /* clear RCVR FIFO           */
+#define FCR_XMITCLR BIT(2)  /* clear XMIT FIFO           */
+#define FCR_FIFO_1  0       /* 1 byte in RCVR FIFO       */
+
 static void serout(int c)
 {
-	while ((INBYTE(REG_LSR(base)) & LSR_THRE) == 0) {
+	while ((IN(REG_LSR) & LSR_THRE) == 0) {
 	}
-	OUTBYTE(REG_THR(base), c);
+	OUT(REG_THR, c);
 }
 
-static int console_out(int c)
+int arch_printk_char_out(int c)
```

Comment:
If we're adding a new arch interface, this must be prototyped and documented in the arch interface headers.

I might be missing something obvious (was there a patch that got dropped?) but nothing in this PR references this function so I don't see how this works. I prefer that we left this and the invocation of `__printk_hook_install()` as-is.

Reply:
Heyyyy I'm just dumb...arch_printk_char_out() infrastructure is already in place. disregard! this is better.